src/app_config.rs

```diff
@@ -0,0 +1,59 @@
+use anyhow::{anyhow, Result};
+use serde::Deserialize;
+use std::{
+    collections::HashMap,
+    fs,
+    path::{Path, PathBuf},
+};
+
+pub(crate) const CONFIG_FILE_NAME: &str = "cargo-generate";
+
+#[derive(Deserialize)]
+pub(crate) struct AppConfig {
+    pub favorites: HashMap<String, FavoriteConfig>,
+}
+
+#[derive(Deserialize)]
+pub(crate) struct FavoriteConfig {
+    pub description: Option<String>,
+    pub git: Option<String>,
+    pub branch: Option<String>,
+}
+
+impl Default for AppConfig {
+    fn default() -> Self {
+        Self {
+            favorites: HashMap::new(),
+        }
+    }
+}
+
+impl AppConfig {
+    pub(crate) fn from_path(path: &Path) -> Result<AppConfig> {
+        if path.exists() {
+            Ok(fs::read_to_string(path).and_then(|c| {
+                if c.trim().is_empty() {
+                    Ok(Default::default())
+                } else {
+                    Ok(toml::from_str(&c)?)
+                }
+            })?)
+        } else {
+            Ok(Default::default())
+        }
+    }
+}
+
+pub(crate) fn app_config_path(path: &Option<PathBuf>) -> Result<PathBuf> {
+    Ok(path
```

Comment:
this is still a bit of doublication, since the inner returns a result you would not create a new result with `Ok()` but rather relay on the actions that produces a result. For example:

```rust
pub(crate) fn app_config_path(path: &Option<PathBuf>) -> Result<PathBuf> {
    path
        .clone()
        .or_else(|| home::cargo_home().map_or(None, |h| Some(h.join(CONFIG_FILE_NAME))))
        .ok_or_else(|| {
            anyhow!(
                r#"
Unable to resolve path for configuration file.
Use --config option, or place {} in $CARGO_HOME."#,
                CONFIG_FILE_NAME
            )
        })
}
```

Reply:
but even better would be to use the [`.with_context`](https://docs.rs/anyhow/1.0.38/anyhow/trait.Context.html) method of anyhow something like:
```rust
pub(crate) fn app_config_path(path: &Option<PathBuf>) -> Result<PathBuf> {
    path
        .clone()
        .or_else(|| home::cargo_home().map_or(None, |h| Some(h.join(CONFIG_FILE_NAME))))
        .with_context(||, format!("Unable to resolve path for configuration file.\nUse --config option, or place {} in $CARGO_HOME.",
                CONFIG_FILE_NAME))
}
```

Reply:
It's too bad clippy don't see these :-)

Originally I had it wrapped because I needed the error conversion, that got refactored out but I never caught to rest.
 Didn't know of `with_context`, but hey, can't know of every crate out there - besides, day job is in C#.

I'll fix it!